fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,635 @@
+                }
+                return nextStatsResult;
+            }).whenComplete((ignore, err) -> kvCursor.close()); //no need to keep it open as it must be reconstituted with a continuation later
+        });
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return context.getExecutor();
+    }
+
+    @Nonnull
+    @Override
+    public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+        visitor.visitEnter(this);
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public CompletableFuture<Boolean> onHasNext() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public SizeStatisticsResults next() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nullable
+    @Override
+    @Deprecated
+    public byte[] getContinuation() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Nonnull
+    @Override
+    @Deprecated
+    public RecordCursor.NoNextReason getNoNextReason() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void close() {
+        if (hasNextFuture != null) {
+            hasNextFuture.cancel(false);
+        }
+    }
+
+    //form a continuation that allows us to restart statistics aggregation from where we left off
+    private class SizeStatisticsCollectorCursorContinuation implements RecordCursorContinuation {
+        @Nonnull
+        private final RecordCursorResult<KeyValue> currentKvResult;
+        @Nonnull
+        private final Function<byte[], RecordCursorProto.SizeStatisticsContinuation> continuationFunction;
+        @Nonnull
+        private final SizeStatisticsResults sizeStatisticsResults;
+        @Nullable
+        private byte[] cachedBytes;
+
+        private SizeStatisticsCollectorCursorContinuation(RecordCursorResult<KeyValue> currentKvResult, SizeStatisticsResults sizeStatisticsResults) {
+            this.cachedBytes = null;
+            this.currentKvResult = currentKvResult;
+            this.sizeStatisticsResults = sizeStatisticsResults.copy(); //cache an immutable snapshot of the partial aggregate state
+
+            //defer forming bytes until requested
+            continuationFunction = b -> {
+                RecordCursorProto.SizeStatisticsPartialResults.Builder pr;
+                if (finalResultsEmitted == false) {
+                    //more to aggregate; put partial results in continuation
+                    pr = RecordCursorProto.SizeStatisticsPartialResults.newBuilder().setKeyCount(this.sizeStatisticsResults.getKeyCount()).setKeySize(this.sizeStatisticsResults.getKeySize()).setMaxKeySize(this.sizeStatisticsResults.getMaxKeySize()).setValueSize(this.sizeStatisticsResults.getValueSize()).setMaxValueSize(this.sizeStatisticsResults.getMaxValueSize());
```

Comment:
Should this be encapsulated in a `toProto` method on `SizeStatisticsResults`?

Reply:
It is. See line 103 of SizeStatisticsCollectorCursor. 
this.sizeStatisticsResults.setFromProto(statsContinuation.getPartialResults());

Reply:
That's the `fromProto` method, sure. But this is about going the other way and producing the protobuf message from the `SizeStatisticsResults`.

Reply:
I think this is still outstanding.